allennlp/models/encoder_decoders/wikitables_semantic_parser.py

```diff
@@ -271,6 +310,166 @@ def forward(self,  # type: ignore
+        entity_types : ``Dict[int,str]``
+            This is a mapping from (batch_index, entity_index) to entity type id.
+        """
+        entity_types = {}
+        batch_types = []
+        for batch_index, world in enumerate(worlds):
+            types = []
+            for entity_index, entity in enumerate(world.table_graph.entities):
+                cell_type_one_hot = [1, 0]
+                row_type_one_hot = [0, 1]
+                types.append((cell_type_one_hot if entity.startswith('fb:cell') else row_type_one_hot))
+
+                # For easier lookups later, we're actually using a _flattened_ version
+                # of (batch_index, entity_index) for the key, because this is how the
+                # linking scores are stored.
+                flattened_entity_index = batch_index * num_entities + entity_index
+                entity_type = 0 if entity.startswith('fb:cell') else 1
+                entity_types[flattened_entity_index] = entity_type
+            padded = pad_sequence_to_length(types, num_entities, lambda: [0, 0])
+            batch_types.append(padded)
+        return Variable(tensor.data.new(batch_types)), entity_types
+
+    def _get_prob_entity(self,
+                         worlds: List[WikiTablesWorld],
+                         scores: Variable,
+                         batch_size: int,
+                         num_entities: int,
+                         num_question_tokens: int,
+                         tensor: torch.Tensor) -> Variable:
+        """
+        Produces the probability of an entity given a question word and type. The logic below
+        separates the entities by type since the softmax normalization term sums over entities
+        of a single type.
+
+        Parameters
+        ----------
+        worlds : ``List[WikiTablesWorld]``
+        scores : ``torch.autograd.Variable``
+            Has shape (batch_size, num_entities, num_question_tokens).
+        batch_size : ``int``
+        num_entities : ``int``
+        num_question_tokens : ``int``
+        tensor : ``torch.Tensor``
+            Used for copying the constructed list onto the right device.
+
+        Returns
+        -------
+        batch_probabilities : ``torch.autograd.Variable``
+            Has shape ``(batch_size, num_entities+1, num_question_tokens)``.
+            Contains all the probabilities for an entity given a question word.
+        """
+        batch_probabilities = Variable(tensor.data.new(torch.zeros(batch_size,
+                                                                   num_entities+1,
+                                                                   num_question_tokens)))
+        for batch_index, world in enumerate(worlds):
+            cell_type_index, row_type_index = self._get_entity_index_by_type(world)
+
+            # Separate the scores by type, since normalization summed over types.
+            cell_type_scores = torch.index_select(scores[batch_index],
```

Comment:
With all of these slices, `index_selects`, and element assignments, this is going to be _really_ slow.

...

Well, I was thinking about ways to fix this to make it faster, but they are pretty complicated.  How about we profile this to make sure it's actually slow before making those complicated changes?  Have you tried training this?